File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/java_to_py/StockPortfolioTracker.py
class StockPortfolioTracker:
    class Stock:
        def __init__(self, name, price, quantity):
            self.name = name
            self.price = price
            self.quantity = quantity

        def get_name(self):
            return self.name

        def get_price(self):
            return self.price

        def get_quantity(self):
            return self.quantity

        def set_quantity(self, quantity):
            self.quantity = quantity

        def __eq__(self, other):
            if not isinstance(other, Stock):
                return False
            return (self.name == other.name and 
                    self.price == other.price and 
                    self.quantity == other.quantity)

        def __hash__(self):
            return hash((self.name, self.price, self.quantity))

    def __init__(self, initial_cash_balance):
        self.initial_cash_balance = initial_cash_balance
        self.cash_balance = initial_cash_balance
        self.portfolio = []

    def add_stock(self, stock):
        for s in self.portfolio:
            if s.get_name() == stock.get_name() and s.get_price() == stock.get_price():
                s.set_quantity(s.get_quantity() + stock.get_quantity())
                return
        self.portfolio.append(stock)

    def remove_stock(self, stock):
        for s in self.portfolio:
            if s.get_name() == stock.get_name() and s.get_price() == stock.get_price():
                if s.get_quantity() >= stock.get_quantity():
                    s.set_quantity(s.get_quantity() - stock.get_quantity())
                    if s.get_quantity() == 0:
                        self.portfolio.remove(s)
                    return True
        return False

    def buy_stock(self, stock):
        cost = stock.get_price() * stock.get_quantity()
        if self.cash_balance >= cost:
            self.add_stock(stock)
            self.cash_balance -= cost
            return True
        return False

    def sell_stock(self, stock):
        revenue = stock.get_price() * stock.get_quantity()
        if self.remove_stock(stock):
            self.cash_balance += revenue
            return True
        return False

    def get_portfolio(self):
        return [Stock(s.name, s.price, s.quantity) for s in self.portfolio]

    def get_cash_balance(self):
        return self.cash_balance

    def calculate_portfolio_value(self):
        total_value = 0.0
        for stock in self.portfolio:
            total_value += stock.get_price() * stock.get_quantity()
        return total_value

    def get_portfolio_summary(self):
        summary = []
        for stock in self.portfolio:
            summary.append(f"{stock.get_name()}: {stock.get_quantity()} shares at ${stock.get_price()} each")
        total_value = self.calculate_portfolio_value()
        summary.append(f"Total Value: ${total_value}")
        return "\n".join(summary)

File: data/derived_cot/rq1_traces/translation/r1/java_to_cpp/code_output/StockPortfolioTracker.cpp
#include <vector>
#include <string>
#include <cmath>
#include <cstring>
#include <cstdint>
#include <sstream>
#include <iostream>

class StockPortfolioTracker {
public:
    class Stock {
    public:
        Stock(const std::string& name, double price, int quantity)
            : name(name), price(price), quantity(quantity) {}

        const std::string& getName() const { return name; }
        double getPrice() const { return price; }
        int getQuantity() const { return quantity; }
        void setQuantity(int newQuantity) { quantity = newQuantity; }

        bool operator==(const Stock& other) const {
            return (name == other.name) &&
                   (doubleToLongBits(price) == doubleToLongBits(other.price)) &&
                   (quantity == other.quantity);
        }

        std::string toString() const {
            std::ostringstream oss;
            oss << name << ": " << quantity << " shares at $" << price << " each";
            return oss.str();
        }

    private:
        std::string name;
        double price;
        int quantity;

        static uint64_t doubleToLongBits(double x) {
            if (std::isnan(x)) {
                return 0x7FF8000000000000;
            }
            uint64_t bits;
            std::memcpy(&bits, &x, sizeof(double));
            return bits;
        }
    };

    StockPortfolioTracker(double initialCashBalance)
        : initialCashBalance(initialCashBalance), cashBalance(initialCashBalance) {}

    void addStock(const Stock& stock) {
        for (auto& s : portfolio) {
            if (s.getName() == stock.getName() && s.getPrice() == stock.getPrice()) {
                s.setQuantity(s.getQuantity() + stock.getQuantity());
                return;
            }
        }
        portfolio.push_back(stock);
    }

    bool removeStock(const Stock& stock) {
        for (auto it = portfolio.begin(); it != portfolio.end(); ++it) {
            if (it->getName() == stock.getName() && it->getPrice() == stock.getPrice()) {
                if (it->getQuantity() >= stock.getQuantity()) {
                    it->setQuantity(it->getQuantity() - stock.getQuantity());
                    if (it->getQuantity() == 0) {
                        portfolio.erase(it);
                    }
                    return true;
                }
            }
        }
        return false;
    }

    bool buyStock(const Stock& stock) {
        double cost = stock.getPrice() * stock.getQuantity();
        if (cashBalance >= cost) {
            addStock(stock);
            cashBalance -= cost;
            return true;
        }
        return false;
    }

    bool sellStock(const Stock& stock) {
        double revenue = stock.getPrice() * stock.getQuantity();
        if (removeStock(stock)) {
            cashBalance += revenue;
            return true;
        }
        return false;
    }

    std::vector<Stock> getPortfolio() const {
        return portfolio;
    }

    double getCashBalance() const {
        return cashBalance;
    }

    double calculatePortfolioValue() const {
        double totalValue = 0.0;
        for (const auto& stock : portfolio) {
            totalValue += stock.getPrice() * stock.getQuantity();
        }
        return totalValue;
    }

    std::string getPortfolioSummary() const {
        std::ostringstream summary;
        for (const auto& stock : portfolio) {
            summary << stock.toString() << "\n";
        }
        summary << "Total Value: $" << calculatePortfolioValue() << "\n";
        return summary.str();
    }

private:
    double initialCashBalance;
    double cashBalance;
    std::vector<Stock> portfolio;
};